# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.h
# Commit: 290fcbe92f0a
# Full Hash: 290fcbe92f0a6d6f5460bc4c374675329f63a496
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-30 04:06:43
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Implement VideoDecoder::Flush r=padenot
#   
#   This patch implements the flush method [1] for WebCodecs' VideoDecoder
#   interface, and add some code that reset method needs.
#   
# ==============================================================================

diff -r 57df271c6d6c -r 290fcbe92f0a dom/media/webcodecs/VideoDecoder.h
--- a/dom/media/webcodecs/VideoDecoder.h	Thu Jun 29 22:33:12 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.h	Thu Jun 29 22:33:13 2023 +0000
@@ -74,6 +74,16 @@
   UniquePtr<ChunkData> mData;
 };
 
+struct FlushMessage final {
+  using Id = size_t;
+
+  FlushMessage(Id aId, Promise* aPromise);
+  const nsCString mTitle;  // Used to identify the message in the logs.
+  MozPromiseRequestHolder<DecoderAgent::DecodePromise> mRequest;
+  const Id mId;  // A unique id shown in log.
+  RefPtr<Promise> mPromise;
+};
+
 class VideoDecoder final : public DOMEventTargetHelper {
  public:
   NS_DECL_ISUPPORTS_INHERITED
@@ -106,7 +116,7 @@
 
   MOZ_CAN_RUN_SCRIPT void Decode(EncodedVideoChunk& aChunk, ErrorResult& aRv);
 
-  already_AddRefed<Promise> Flush(ErrorResult& aRv);
+  MOZ_CAN_RUN_SCRIPT already_AddRefed<Promise> Flush(ErrorResult& aRv);
 
   void Reset(ErrorResult& aRv);
 
@@ -135,10 +145,13 @@
 
   void ScheduleDequeueEvent();
 
+  void SchedulePromiseResolveOrReject(Promise* aPromise,
+                                      const nsresult& aResult);
+
   MOZ_CAN_RUN_SCRIPT void ProcessControlMessageQueue();
-  void CancelPendingControlMessages();
+  void CancelPendingControlMessages(const nsresult& aResult);
 
-  using ControlMessage = Variant<ConfigureMessage, DecodeMessage>;
+  using ControlMessage = Variant<ConfigureMessage, DecodeMessage, FlushMessage>;
   enum class MessageProcessedResult { NotProcessed, Processed };
 
   MOZ_CAN_RUN_SCRIPT MessageProcessedResult
@@ -146,6 +159,8 @@
 
   MessageProcessedResult ProcessDecodeMessage(ControlMessage& aMessage);
 
+  MessageProcessedResult ProcessFlushMessage(ControlMessage& aMessage);
+
   // Returns true when mAgent can be created.
   bool CreateDecoderAgent(UniquePtr<VideoDecoderConfig>&& aConfig,
                           UniquePtr<TrackInfo>&& aInfo);
@@ -171,6 +186,7 @@
   bool mDequeueEventScheduled;
 
   DecodeMessage::Id mDecodeMessageCounter;
+  FlushMessage::Id mFlushMessageCounter;
 
   // Used to add a nsIAsyncShutdownBlocker on main thread to block
   // xpcom-shutdown before the underlying MediaDataDecoder is created. The
